### PR TITLE
[hail] Makefile

### PR DESCRIPTION
**Pull request by danking**
- move environment.yml files out of the packaged directory so they don't get shipped to end users.

 - add `make test-pip-deploy` which pip deploys to the next available `devN` version (you have to wait a bit before you can `pip install` it, so I didn't include that in the test, but you can do that manually, or a motivated person can write a polling script)

 - add `build/dev-conda` which ensures that if the dev-environment file changes since you last ran `make build/dev-conda`, your conda environment is updated

 - pedantically use the correct conda environment _everywhere_

 - use python to determine cpu count instead of fixing it at 2

 - add `jq` as an `env-setup.sh` dependency

 - add `make build/credentials.json` which `scp`s a new JSON file containing credentials to the local machine, moreover there are two rules for automatically extracting the credentials for PYPI from this JSON file

 - use `ENV_VAR`, a make macro, to ensure we rebuild the appropriate targets (but no more) when a relevant environment variable is changed since last build

 - added several missing breeze versions, now we can easily test against new spark versions, just run `SPARK_VERSION=4.0.0 make test`

 - fold doctests in with regular tests under `test-python` which uses pytest, no more unnecessary copying as well

 - fix build-info

 - delete two unused python files in hail root

 - correct LIBSIMDPP dependency in C makefile

# Not Doing Yet

 - incorporate native lib into this Makefile. Instead, if anything changed in src/main/c since we last built, we rebuild

 - fix the directory structure to be compliant with pytests recommended structure

### CODE REVIEW

**Comment by danking** [comment]
replaces https://github.com/hail-is/hail/pull/5130

**Comment by danking** [comment]
```
(base) dking@wmb16-359 # time make jar
make: Nothing to be done for `jar'.
make jar  0.05s user 0.07s system 83% cpu 0.147 total
```

File: hail/Makefile [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+GIT_SHA := $(shell git rev-parse HEAD)
+
+$(eval $(call ENV_VAR,SPARK_VERSION))
+$(eval $(call ENV_VAR,HAIL_PIP_VERSION))
+$(eval $(call ENV_VAR,GRADLE_ARGS))
+$(eval $(call ENV_VAR,GIT_SHA))
+
+IGNORED_FILE_PATTERN := \( -name '*~' \)
+JVM_FILES := $(shell find src/main/java src/main/scala src/main/resources \( -not $(IGNORED_FILE_PATTERN) \))
+IGNORED_PY_FILE_PATTERN = \( $(IGNORED_FILE_PATTERN) -o -name '*.pyc' \)
```
Comment (tpoterba) [inline hail/Makefile]: `hail-*.log`?

Maybe we should add a task to delete log files?

File: hail/Makefile [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+MAKEFLAGS += --no-builtin-rules
+.SUFFIXES:
+
+PARALLELISM ?= $(shell python -c 'import multiprocessing; print(multiprocessing.cpu_count())')
+SPARK_VERSION ?= 2.2.0
+# Hail has three notions of version:
+#  - hail short version: MAJOR.MINOR
+#  - hail pip version: MAJOR.MINOR.PATCH
+#  - hail version: MAJOR.MINOR.PATCH-GIT_SHA, calculated by generateBuildInfo
+MAJOR ?= 0
```
[the remark this answers is not in the document]
Comment (danking) [reply hail/Makefile]: good point.

File: hail/Makefile (hunk 1 above)
Comment (danking) [reply hail/Makefile]: i didn't see any files in our hierarchy that end in `.log` so I ignored all `.log` files. Also added this to the .gitignore.

Comment (danking) [reply hail/Makefile]: I like having a task to delete log files. Wanna PR that after this one lands?

Comment (tpoterba) [reply hail/Makefile]: sure, sounds good.

**Comment by danking** [comment]
@tpoterba good point, damn it feels so good to delete docs. https://github.com/hail-is/hail/blob/469e74dcb7a40dcaaa4d4d384c4d2dbebfc809e6/hail/python/hail/docs/getting_started_developing.rst

**Comment by tpoterba** [comment]
Great change. Still an error related to hail_pip_version in the tests, though.

Can you make a discuss post when this goes in to alert people compiling their own builds?

**Comment by danking** [comment]
I feel a bit like a cheat here since there's been a fair bit of work since you last reviewed. Most of it was fixes of tiny bugs that the CI revealed. There was [one, kind of notable, change](https://github.com/hail-is/hail/pull/5194/commits/f95e4e0ff1cdd2865cf703aa27f780c7f162316c). I removed Spark from the Dockerfile. It is no longer necessary because the pip install will pull the correct version of Spark. To avoid pulling Spark on each PR build, I cache 2.2.0 (and all our other pip dependencies) in the hail conda env in the docker image. Doing this required that I move our requirements into a requirements file which is parameterized by spark version.

A good follow up PR would be to either a) entirely remove dependency on conda or b) generate the conda `environment.yml` from `requirements.txt.in`.

File: hail/hail-ci-build.sh [hunk 3]
```diff
@@ -204,29 +196,27 @@ export GRADLE_OPTS="-Xmx2048m"
 export GRADLE_USER_HOME="/gradle-cache"
 
 echo "Compiling..."
-./gradlew shadowJar archiveZip > ${COMPILE_LOG} 2>&1
+make jar zip
```
Comment (tpoterba) [inline hail/hail-ci-build.sh]: Put back the redirection?

Comment (danking) [reply hail/hail-ci-build.sh]: whoops good catch.

**Comment by danking** [comment]
sorry I'll come back to this, the doctests are broke somehow.

**Pull request closed without merging**